File: core/sds-aspect-meta-model-java/src/main/java/io/openmanufacturing/sds/metamodel/impl/DefaultEntityInstance.java
/*
 * Copyright (c) 2022 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package io.openmanufacturing.sds.metamodel.impl;

import java.util.Map;
import java.util.Objects;
import java.util.StringJoiner;

import io.openmanufacturing.sds.metamodel.Entity;
import io.openmanufacturing.sds.metamodel.EntityInstance;
import io.openmanufacturing.sds.metamodel.Property;
import io.openmanufacturing.sds.metamodel.Value;
import io.openmanufacturing.sds.metamodel.loader.MetaModelBaseAttributes;
import io.openmanufacturing.sds.metamodel.visitor.AspectVisitor;

public class DefaultEntityInstance extends BaseImpl implements EntityInstance {
   private final Map<Property, Value> assertions;
   private final Entity type;

   public DefaultEntityInstance( final MetaModelBaseAttributes metaModelBaseAttributes, final Map<Property, Value> assertions, final Entity type ) {
      super( metaModelBaseAttributes );
      this.assertions = assertions;
      this.type = type;
   }

   @Override
   public Entity getType() {
      return type;
   }

   @Override
   public Map<Property, Value> getAssertions() {
      return assertions;
   }

   @Override
   public int compareTo( final EntityInstance other ) {
      return getName().compareTo( other.getName() );
   }

   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitEntityInstance( this, context );
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultEntityInstance.class.getSimpleName() + "[", "]" )
            .add( "assertions=" + assertions )
            .add( "type=" + type )
            .toString();
   }

   @Override
   public boolean equals( final Object o ) {
      if ( this == o ) {
         return true;
      }
      if ( o == null || getClass() != o.getClass() ) {
         return false;
      }
      if ( !super.equals( o ) ) {
         return false;
      }
      final DefaultEntityInstance that = (DefaultEntityInstance) o;
      return Objects.equals( assertions, that.assertions ) && Objects.equals( type, that.type );
   }

   @Override
   public int hashCode() {
      return Objects.hash( super.hashCode(), assertions, type );
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/esmf/aspect-meta-model/default-entity-instance.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import {Entity} from './default-entity';
import {IsInstance} from './is-instance';

export class DefaultEntityInstance implements IsInstance<Entity> {
    private _metaModelType: Entity;
    private _name: string;
    private _descriptions: Map<string, string>;

    constructor(name: string, metaModelType: Entity, description?: Map<string, string>) {
        this._name = name;
        this._metaModelType = metaModelType;
        this._descriptions = description || new Map();
    }

    public get metaModelType(): Entity {
        return this._metaModelType;
    }

    public get name(): string {
        return this._name;
    }

    public get descriptionKey(): string | undefined {
        return this._metaModelType.properties.find(
            property => property.name.toLowerCase().includes('description') && property.isNotInPayload === true
        ).name;
    }

    public get valuePayloadKey(): string {
        return this._metaModelType.properties.find(property => property.isNotInPayload === false).name;
    }

    public get value(): string | number | boolean {
        return this[this.valuePayloadKey];
    }

    public getDescription(locale = 'en'): string | undefined {
        return this._descriptions.get(locale);
    }

    public get localesDescriptions(): Array<string> {
        return Array.from(this._descriptions.keys());
    }
}
